✨Support specific kubeconfig context in clusterctl

wfernandes opened the pull request:
<!-- please add a icon to the title of this PR (see https://sigs.k8s.io/cluster-api/VERSIONING.md), and delete this line and similar ones -->
<!-- the icon will be either ⚠️ (:warning:, major or breaking changes), ✨ (:sparkles:, minor or feature additions), 🐛 (:bug:, patch and bugfixes), 📖 (:book:, documentation or proposals), or 🏃 (:running:, other) -->

**What this PR does / why we need it**:
This PR adds support to specify the context within a kubeconfig file. It affects all the commands that provide `--kubeconfig` flag.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #2659 
Fixes #2854

**Release Notes:**
- 🐛  Fixes bug in `proxy.CurrentNamespace` to return error if it cannot find
context in the kubeconfig.
- ➕ Adds GetConfig to Proxy interface
- ➕ Adds alias client.Kubeconfig to cluster.Kubeconfig to continue separation
of concerns between client libraries.
- ➕ ClusterClientFactory takes a new input called client.Kubeconfig
which contains kubeconfig Path and Context.
- ⚠️ Replaces KubeconfigPath() method on cluster.Client interface with Kubeconfig().
  - This method returns cluster.Kubeconfig which contains the various kubeconfig properties that are supported.
- ➕ Adds `--kubeconfig-context` flag to `move` command.
  - Replace FromKubeconfigPath in client.MoveOptions to FromKubeconfig.
- ➕ Adds `--to-kubeconfig-context` flag to `move` command.
  - Replace ToKubeconfigPath in client.MoveOptions to ToKubeconfig.
- ➕ Adds `--kubeconfig-context` flag to `config cluster` command.
  - Replace KubeconfigPath in client.GetClusterTemplateOptions to Kubeconfig.
- ➕ Adds `--kubeconfig-context` flag to `delete` command.
  - Replace KubeconfigPath in client.DeleteOptions to Kubeconfig.
- ➕ Adds `--kubeconfig-context` flag to `init` command.
  - Replace KubeconfigPath in client.InitOptions to Kubeconfig.
- ➕ Adds `--kubeconfig-context` flag to `upgrade` command.
  - Replace KubeconfigPath in client.PlanUpgradeOptions to Kubeconfig.
  - Replace KubeconfigPath in client.ApplyUpgradeOptions to Kubeconfig.
- Override timeout in proxy client
  - Defaults to a 30s timeout. This can be overridden via the InjectProxyTimeout ProxyOption.

wfernandes commented:
To those considering reviewing this PR, try looking at the changes commit by commit. I tried to break it up by logical commits to make sense of what is going on.

Also I've added **basic unit tests** for some areas that didn't have unit tests (`InitImages`, `ProxyGetConfig`, `Move`, `PlanUpgrade`). We'll need to revisit them to determine if more test cases are needed or not. For example, I believe there is a whole e2e dedicated for `Move`.

wfernandes commented:
Performing a manual test for some of the commands.

benmoss commented:
I think instead of adding another getter to the `Client` interface we could replace `Kubeconfig() string` with a client-go `api.Config`. That way clients don't have to write files to disk in order to use it. We can have some helper constructor to do that as an easier API for the CLI to use, but as an API I think it's better to accept a struct.

This will also simplify a lot of the tests so we don't have to write files to disk to have a kubeconfig.

wfernandes commented:
@benmoss I did think about using that struct but I was trying to minimize the amount of changes to the API. We had a discussion and decided that adding an option for context would be fine.

But if we don't mind further changes, I can see that the `clusterClient` could load the kubeconfig file in `newClusterClient`  and hold on to the reference of [`clientcmdapi.Config`](https://github.com/kubernetes/client-go/blob/cb664d40f84c27bee45c193e4acb0fcd549b0305/tools/clientcmd/api/types.go#L31) and then just pass that into `newProxy` instead of the `kubeconfigPath` and `kubeconfigContext`.

However, that would mean some more changes to the `clusterClientFactory` or preferably just remove the usage of factories all together. We could have the `clusterctlClient` reference a `cluster.Client` instead of a `ClusterClientFactory`.

So yeah there are options 🙂 
@fabriziopandini If you have input on the need for the `ClusterClientFactory` I can switch to using the [`clientcmdapi.Config`](https://github.com/kubernetes/client-go/blob/cb664d40f84c27bee45c193e4acb0fcd549b0305/tools/clientcmd/api/types.go#L31).

fabriziopandini commented on cmd/clusterctl/client/cluster/proxy.go:
```diff
@@ -89,8 +95,35 @@ func (k *proxy) ValidateKubernetesVersion() error {
+		return nil, errors.Wrapf(err, "failed to load Kubeconfig file from %q", k.kubeconfigPath)
+	}
+
+	restConfig, err := clientcmd.NewNonInteractiveClientConfig(
```
Is there a reason to switch to `NewNonInteractiveClientConfig` instead of  using `NewDefaultClientConfig` as before and just set the `Contect` field in the `ConfigOverrides{}` struct?

fabriziopandini replied:
Aside note. Should we set the timeout on the connection as well? If I'm not wrong no timeout means infinite, and this can lead to weird errors
@vincepri opinions ^^

vincepri replied:
I'm fine either way, I think when @wfernandes and I looked at it, it wasn't clear you could override the configuration to set the context

wfernandes replied:
Whoa! Didn't see that in the `ConfigOverrides` struct. When I saw the definition of `NewDefaultClientConfig` my eyes were immediately drawn to the function argument. Nice...I'll try it with the overrides and make sure everything works as expected.

wfernandes replied:
@fabriziopandini I forgot about your comment regarding timeout. I'll add support to configure that and put in a default timeout. I'll set the default to be 30s unless we have a consistent timeout set for api-server requests in other places.

wfernandes commented:
This PR is ready for review. I'm going to place a hold so that if it looks good I have a chance to squash all the commits into one which will document all the changes.
/hold

vincepri commented on cmd/clusterctl/client/cluster/proxy.go:
```diff
@@ -89,8 +95,35 @@ func (k *proxy) ValidateKubernetesVersion() error {
 	return nil
 }
 
+func (k *proxy) GetConfig() (*rest.Config, error) {
```
I suggested to make it public in the interface so it'd be testable from outside, or you can fake it. In general, I can see some use cases where they want to get the rest config directly

fabriziopandini replied:
thanks for explanation!

vincepri commented:
/milestone v0.3.4

fabriziopandini commented on cmd/clusterctl/cmd/delete.go:
```diff
@@ -84,6 +85,8 @@ var deleteCmd = &cobra.Command{
 	deleteCmd.Flags().StringVar(&dd.kubeconfig, "kubeconfig", "",
 		"Path to the kubeconfig file to use for accessing the management cluster. If unspecified, default discovery rules apply.")
+	deleteCmd.Flags().StringVar(&dd.kubeconfigContext, "kubeconfig-context", "",
+		"Context corresponding to the management cluster within the kubeconfig file. If empty, default discovery rules apply.")
```
```suggestion
		"Context to be used within the kubeconfig file. If empty, current context will be used.")
```

fabriziopandini commented on cmd/clusterctl/cmd/init.go:
```diff
@@ -91,6 +92,8 @@ var initCmd = &cobra.Command{
 	initCmd.Flags().StringVar(&io.kubeconfig, "kubeconfig", "",
 		"Path to the kubeconfig for the management cluster. If unspecified, default discovery rules apply.")
+	initCmd.Flags().StringVar(&io.kubeconfigContext, "kubeconfig-context", "",
+		"Context corresponding to the management cluster within the kubeconfig file. If empty, default discovery rules apply.")
```
```suggestion
		"Context to be used within the kubeconfig file. If empty, current context will be used.")
```

fabriziopandini commented on cmd/clusterctl/cmd/move.go:
```diff
@@ -52,6 +54,10 @@ func init() {
 	moveCmd.Flags().StringVar(&mo.toKubeconfig, "to-kubeconfig", "",
 		"Path to the kubeconfig file to use for the destination management cluster.")
+	moveCmd.Flags().StringVar(&mo.fromKubeconfigContext, "kubeconfig-context", "",
+		"Context corresponding to the source management cluster within the kubeconfig file. If empty, default discovery rules apply.")
```
```suggestion
		"Context to be used within the kubeconfig file for the source management cluster. If empty, current context will be used.")
```

fabriziopandini commented on cmd/clusterctl/cmd/move.go:
```diff
@@ -52,6 +54,10 @@ func init() {
+	moveCmd.Flags().StringVar(&mo.fromKubeconfigContext, "kubeconfig-context", "",
+		"Context corresponding to the source management cluster within the kubeconfig file. If empty, default discovery rules apply.")
+	moveCmd.Flags().StringVar(&mo.toKubeconfigContext, "to-kubeconfig-context", "",
+		"Context corresponding to the destination management cluster within the to-kubeconfig file. If empty, default discovery rules apply.")
```
```suggestion
		"Context to be used within the kubeconfig file for the destination management cluster. If empty, current context will be used.")
```

wfernandes commented on cmd/clusterctl/client/client_test.go:
```diff
@@ -195,18 +200,24 @@ func (p *fakeCertManagerClient) Images() ([]string, error) {
+	kubeconfig        string
+	kubeconfigContext string
+	fakeProxy         *test.FakeProxy
+	fakeObjectMover   cluster.ObjectMover
```
I added a `client/move_test.go` with some basic unit tests which references it.

wfernandes commented on cmd/clusterctl/client/cluster/proxy_test.go:
```diff
@@ -0,0 +1,204 @@
+	tests := []struct {
+		name               string
+		context            string
+		currentContext     string
```
Nope. This is a field that sets the `currentContext: ` in the kubeconfig. I specify this field to ensure that if the `currentContext` is something different from the `context` that is passed in it respects the `context` and not `currentContext`.

The assertion is indirectly happening on the host (`expectedHost`) of the corresponding cluster to the context since that's what's available from the `rest.Config` that is returned.

wfernandes commented on cmd/clusterctl/client/cluster/proxy_test.go:
```diff
@@ -0,0 +1,204 @@
+		{
+			name:               "returns error if cannot load the kubeconfig file",
+			expectErr:          true,
+			kubeconfigContents: "bad contents",
```
Sure. I can do that if it improves readability. 🙂

wfernandes commented on cmd/clusterctl/client/cluster/proxy.go:
```diff
@@ -38,24 +38,30 @@ var (
-	if config.CurrentContext == "" {
-		return "", errors.Wrapf(err, "failed to get current-context from %q", k.kubeconfig)
+	context := config.CurrentContext
+	if context == "" {
```
Ok. I think I understand. So you are saying let's remove the error `failed to get current-context` altogether? Yeah, that makes sense. So if `context` is `""`, we'll fall back onto the error `failed to get context "" from ...`

wfernandes commented on cmd/clusterctl/client/client.go:
```diff
@@ -110,8 +64,18 @@ type clusterctlClient struct {
 
+// ClusterClientFactoryInput specify the inputs required for creating cluster
+// clients
+type ClusterClientFactoryInput struct {
```
For now, I made `clusterClientFactoryInput` private since it didn't need to be exported.

wfernandes commented on cmd/clusterctl/client/init_test.go:
```diff
@@ -28,6 +28,74 @@ import (
 	"sigs.k8s.io/cluster-api/cmd/clusterctl/internal/util"
 )
 
+// TODO: Refactor to actually test InitImages
```
I created this issue to track that work: https://github.com/kubernetes-sigs/cluster-api/issues/2871

wfernandes commented:
I'm planning on squashing the commits sometime tomorrow. If anyone wants I'm happy to walk through the changes made in this PR.

wfernandes commented:
/remove hold

wfernandes commented:
/hold cancel

vincepri commented:
Code changes LGTM, over to you @fabriziopandini

wfernandes commented:
@vincepri I squashed all the commits and added all the changelog notes to the commit message for the release notes. Let me know if that works or if there is another place you'd like me to add the changelog for this PR.

vincepri commented:
Going through the code a little bit more, one thing I've noticed is that we're using/passing `KubeconfigPath` and `KubeconfigContext` in a lot of places. This is a good short term solution, although we might want to switch these to be our own `Kubeconfig` struct. In the future, we might want to add other Kubeconfig options (e.g. user, etc), having a single struct will greatly simplify future changes.

wdyt?

wfernandes commented:
Definitely! I had suggested something like this before https://github.com/kubernetes-sigs/cluster-api/pull/2856#discussion_r404357373

But I definitely agree. If you'd like we can create a separate issue for that and I can make those changes in a separate PR.
Or I could do it in this PR if that's what you are hinting to 😉

vincepri commented:
I'd prefer to include those in this PR, so we limit the number of times we change up the public library on master

wfernandes commented:
/hold
Adding a hold so that I can get the above changes regarding the `Kubeconfig` struct in this PR.

wfernandes commented:
/hold cancel
@vincepri I added a struct called `cluster.Kubeconfig` and passed that through instead of the separate vars - path, context. This should allow for further expansion on kubeconfig properties if needed in the future.

fabriziopandini commented on cmd/clusterctl/client/client_test.go:
```diff
@@ -195,18 +200,24 @@ func (p *fakeCertManagerClient) Images() ([]string, error) {
-	fakeProxy      *test.FakeProxy
-	repositories   map[string]repository.Client
-	internalclient cluster.Client
+	kubeconfig        string
```
use cluster.Kubeconfig here?

fabriziopandini commented on cmd/clusterctl/client/cluster/client.go:
```diff
@@ -38,14 +38,27 @@ var (
-	// Kubeconfig return the path to kubeconfig used to access to a management cluster.
-	Kubeconfig() string
+	// KubeconfigPath return the path to kubeconfig used to access to a management cluster.
+	KubeconfigPath() string
```
```suggestion
	Kubeconfig() Kubeconfig
```
So we avoid adding a getter whenever the struct changes?

wfernandes replied:
Ahh yes. I wanted to do this. Thanks for bringing this up.

fabriziopandini commented on cmd/clusterctl/client/config.go:
```diff
@@ -51,8 +52,12 @@ func (c *clusterctlClient) GetProviderComponents(provider string, providerType c
-	// Kubeconfig file to use for accessing the management cluster. If empty, default discovery rules apply.
-	Kubeconfig string
+	// KubeconfigPath defines the path to the kubeconfig file to use for accessing the management cluster. If empty, default discovery rules apply.
+	KubeconfigPath string
```
```suggestion
	Kubeconfig cluster.Kubeconfig
```
Might be we should alias the structure as well so we can avoid exposing the cluster package

fabriziopandini commented on cmd/clusterctl/client/delete.go:
```diff
@@ -25,8 +25,12 @@ import (
-	// Kubeconfig file to use for accessing the management cluster. If empty, default discovery rules apply.
-	Kubeconfig string
+	// KubeconfigPath defines the path to the kubeconfig file to use for accessing the management cluster. If empty, default discovery rules apply.
+	KubeconfigPath string
```
What about 
```suggestion
	Kubeconfig cluster.Config
```
(or the cluster.Config alias)

This is user-facing/breaking, but probably more looking forward than the current approach
@vincepri opinions?

same applies for other Options struct in the client package

vincepri replied:
We relaxed the API guarantees (which we should actually document in a README under clusterctl), so +1

wfernandes replied:
I'll map them all to the alias type in the client package and remove the individual properties from the Options across the commands.

vincepri commented:
After this is merged in, let's add a summary of all the public API breaking changes to this PR's description @wfernandes @fabriziopandini

wfernandes commented:
@vincepri Sounds good. I'll add all the changes to the PRs description. I documented most of it in the commit's message so most of the work is done.

wfernandes commented:
Sorry this is taking a while. Seems like I need to make some updates to the `test/framework/clusterctl` 😆 😅 
/hold

wfernandes commented:
/hold cancel

fabriziopandini commented:
/lgtm

vincepri commented:
This is the result from `go-apidiff`:

```
! second, different message for obj type sigs.k8s.io/cluster-api/cmd/clusterctl/client.Kubeconfig struct{Path string; Context string}, part ""
  first:  changed from struct{Path string; Context string} to string
  second: removed
! second, different message for obj type sigs.k8s.io/cluster-api/cmd/clusterctl/client/cluster.Kubeconfig struct{Path string; Context string}, part ""
  first:  changed from struct{Path string; Context string} to string
  second: removed

sigs.k8s.io/cluster-api/test/framework/clusterctl
  Incompatible changes:
  - ConfigClusterInput.Kubeconfig: removed
  - InitInput.Kubeconfig: removed
  - MoveInput.FromKubeconfig: removed
  - MoveInput.ToKubeconfig: removed
  Compatible changes:
  - ConfigClusterInput.KubeconfigPath: added
  - InitInput.KubeconfigPath: added
  - MoveInput.FromKubeconfigPath: added
  - MoveInput.ToKubeconfigPath: added

sigs.k8s.io/cluster-api/test/framework
  Incompatible changes:
  - ClusterProxy: removed
  - NewClusterProxy: removed
  - WatchControllerLogs: removed
  - WatchControllerLogsInput: removed
  - WatchNamespaceEventsInput.LogFolder: removed
  Compatible changes:
  - WatchNamespaceEventsInput.LogPath: added

sigs.k8s.io/cluster-api/test/framework/bootstrap
  Incompatible changes:
  - ClusterProvider: removed
  - CreateKindBootstrapClusterAndLoadImages: removed
  - CreateKindBootstrapClusterAndLoadImagesInput: removed
  - KindClusterOption: removed
  - LoadImagesToKindCluster: removed
  - LoadImagesToKindClusterInput: removed
  - NewKindClusterProvider: removed
  - WithDockerSockMount: removed

sigs.k8s.io/cluster-api/cmd/clusterctl/client
  Incompatible changes:
  - Kubeconfig: removed

sigs.k8s.io/cluster-api/cmd/clusterctl/client/cluster
  Incompatible changes:
  - InjectProxyTimeout: removed
  - Kubeconfig: removed
  - Proxy.GetConfig: removed
  - ProxyOption: removed
```

wfernandes commented:
@vincepri I think some of the diff changes above are from a different commit that made changes to the clusterctl test framework. I'll try and run the go-apidiff tool as well to see if I've missed anything.

wfernandes commented:
Just adding for reference:
```
$ go-apidiff f38fd7d37aace69704b68adea3ee9ad2ef19c6d5 c96fd59faeac7d75d80335048bd681fa4c5f0073

sigs.k8s.io/cluster-api/cmd/clusterctl/client/cluster
  Incompatible changes:
  - (*clusterClient).Kubeconfig: changed from func() string to func() Kubeconfig
  - Client.Kubeconfig: changed from func() string to func() Kubeconfig
  - New: changed from func(string, sigs.k8s.io/cluster-api/cmd/clusterctl/client/config.Client, ...Option) Client to func(Kubeconfig, sigs.k8s.io/cluster-api/cmd/clusterctl/client/config.Client, ...Option) Client
  - Proxy.GetConfig: added

sigs.k8s.io/cluster-api/cmd/clusterctl/client
  Incompatible changes:
  - ApplyUpgradeOptions.Kubeconfig: changed from string to Kubeconfig
  - ClusterClientFactory: changed from func(string) (sigs.k8s.io/cluster-api/cmd/clusterctl/client/cluster.Client, error) to func(Kubeconfig) (sigs.k8s.io/cluster-api/cmd/clusterctl/client/cluster.Client, error)
  - DeleteOptions.Kubeconfig: changed from string to Kubeconfig
  - GetClusterTemplateOptions.Kubeconfig: changed from string to Kubeconfig
  - InitOptions.Kubeconfig: changed from string to Kubeconfig
  - MoveOptions.FromKubeconfig: changed from string to Kubeconfig
  - MoveOptions.ToKubeconfig: changed from string to Kubeconfig
  - PlanUpgradeOptions.Kubeconfig: changed from string to Kubeconfig

sigs.k8s.io/cluster-api/test/framework/clusterctl
  Incompatible changes:
  - ConfigClusterInput.KubeconfigPath: removed
  - InitInput.KubeconfigPath: removed
  - MoveInput.FromKubeconfigPath: removed
  - MoveInput.ToKubeconfigPath: removed
```

vincepri commented:
Thanks!
